freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/SudoRuleService.java

```diff
@@ -0,0 +1,158 @@
+package com.sequenceiq.freeipa.service.freeipa.user;
+
+import static com.sequenceiq.cloudbreak.auth.altus.UmsVirtualGroupRight.ALLOW_PRIVILEGED_OS_OPERATIONS;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import com.google.common.base.Strings;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupRequest;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupService;
+import com.sequenceiq.freeipa.client.FreeIpaClient;
+import com.sequenceiq.freeipa.client.FreeIpaClientException;
+import com.sequenceiq.freeipa.client.FreeIpaClientExceptionUtil;
+import com.sequenceiq.freeipa.client.model.SudoCommand;
+import com.sequenceiq.freeipa.client.model.SudoRule;
+import com.sequenceiq.freeipa.entity.Stack;
+
+@Service
+public class SudoRuleService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SudoRuleService.class);
+
+    private static final String HOST_CATEGORY = "all";
+
+    @Value("${freeipa.sudo.ruleName}")
+    private String ruleName;
+
+    @Value("${freeipa.sudo.allowCommands:}")
+    private Set<String> allowCommands;
+
+    @Value("${freeipa.sudo.denyCommands:}")
+    private Set<String> denyCommands;
+
+    private Set<String> allCommands;
+
+    @Inject
+    private VirtualGroupService virtualGroupService;
+
+    @PostConstruct
+    public void postConstruct() {
+        allCommands = Stream.concat(allowCommands.stream(), denyCommands.stream()).collect(Collectors.toSet());
+    }
+
+    public void setupSudoRule(Stack stack, FreeIpaClient freeIpaClient) throws FreeIpaClientException {
+        if (stack.isAvailable()) {
+            String group = getSudoGroupName(stack.getEnvironmentCrn());
+
+            if (!Strings.isNullOrEmpty(group)) {
+                LOGGER.info("Setup sudo rule '{}' to allow '{}' but deny '{}' commands for user group '{}' on all hosts.",
+                        ruleName, allowCommands, denyCommands, group);
+
+                //It throws FreeIpaClientException in case of missing group.
+                freeIpaClient.groupShow(group);
```

Comment:
so extract this to a method called `validateGroupExistsOnFreeIpa` and you could loose the comment